packages/state-importer/src/main.js

```diff
@@ -0,0 +1,240 @@
+import { createSyncStates } from '@commercetools/sync-actions'
+import pkg from '../package.json'
+import type {
+  ApiConfigOptions,
+  ConstructorOptions,
+  LoggerOptions,
+  StateData,
+  Summary,
+} from '../../../types/state'
+import type { Client, SyncAction } from '../../../types/sdk'
+
+class StateImportError extends Error {
+  error: any
+  summary: string
+  message: string
+
+  constructor(
+    message: string,
+    summary: string = 'No summary provided.',
+    error: any = null
+  ) {
+    super(message)
+
+    this.message = message
+    this.summary = summary
+    this.error = error
+  }
+}
+
+export default class StateImport {
+  // Set flowtype annotations
+  accessToken: string
+  continueOnProblems: boolean
+  client: Client
+  apiConfig: ApiConfigOptions
+  logger: LoggerOptions
+  _summary: Summary
+  syncStates: SyncAction
+
+  constructor(options: ConstructorOptions, logger: LoggerOptions) {
+    if (!options.apiConfig)
+      throw new StateImportError(
+        'The constructor must be passed an `apiConfig` object'
+      )
+
+    this.apiConfig = options.apiConfig
+    this.accessToken = options.accessToken
+    this.continueOnProblems = options.continueOnProblems || false
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          this.accessToken ? `Bearer ${this.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    this.syncStates = createSyncStates()
+
+    this.logger = logger || {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      debug: () => {},
+    }
+    this._resetSummary()
+  }
+
+  static _buildPredicate(states: Array<StateData>) {
+    return `key in ("${states.map(state => state.key).join('", "')}")`
+  }
+
+  // Wrapper function for compatibility with CLI
+  processStream(chunk: Array<StateData>, cb: () => mixed) {
+    this.logger.debug(`Starting conversion of ${chunk.length} states`)
+    return this._processBatches(chunk).then(cb)
+    // No catch block as errors will be caught in the CLI
+  }
+
+  // Public function for direct module usage
+  run(states: Array<StateData>) {
+    return this._processBatches(states)
+  }
+
+  _processBatches(states: Array<StateData>) {
+    const predicate = StateImport._buildPredicate(states)
+    const service = this._createService()
+    const uri = service.where(predicate).build()
+    const existingStatesRequest = { uri, method: 'GET' }
+    return this.client
+      .execute(existingStatesRequest)
+      .then(({ body: { results: existingStates } }: Object) =>
+        this._createOrUpdate(states, existingStates)
+      )
+      .then(() => Promise.resolve())
```

Comment:
Err, why are we doing this?

Reply:
It's a hack to make the CLI work. Without this, the CLI fails even on successful import. I'm still investigating why, but so far, I've been unsuccessful